runtime/compact.c

```diff
@@ -35,95 +35,73 @@
 */
-#ifdef WITH_PROFINFO
-#define Make_ehd(s,t,c,p) \
-  (((s) << 10) | (t) << 2 | (c) | ((p) << PROFINFO_SHIFT))
-#else
-#define Make_ehd(s,t,c,p) (((s) << 10) | (t) << 2 | (c))
-#endif
-#define Whsize_ehd(h) Whsize_hd (h)
-#define Wosize_ehd(h) Wosize_hd (h)
-#define Tag_ehd(h) (((h) >> 2) & 0xFF)
-#ifdef WITH_PROFINFO
-#define Profinfo_ehd(hd) Profinfo_hd(hd)
-#endif
-#define Ecolor(w) ((w) & 3)
 
 typedef uintnat word;
 
+#define eptr(p) \
+  (((word) (p) & ~0x3FF) | ((((word) p) & 0x3FF) >> 2) | Caml_gray)
+#define dptr(p) ((word *) (((word) (p) & ~0x3FF) | ((((word) p) & 0xFF) << 2)))
+
 static void invert_pointer_at (word *p)
 {
   word q = *p;
-  CAMLassert (Ecolor ((intnat) p) == 0);
-
-  /* Use Ecolor (q) == 0 instead of Is_block (q) because q could be an
-     inverted pointer for an infix header (with Ecolor == 2). */
-  if (Ecolor (q) == 0 && Is_in_heap (q)){
-    switch (Ecolor (Hd_val (q))){
-    case 0:
-    case 3: /* Pointer or header: insert in inverted list. */
-      *p = Hd_val (q);
-      Hd_val (q) = (header_t) p;
-      break;
-    case 1: /* Infix header: make inverted infix list. */
-      /* Double inversion: the last of the inverted infix list points to
-         the next infix header in this block.  The last of the last list
-         contains the original block header. */
-      {
-        /* This block as a value. */
-        value val = (value) q - Infix_offset_val (q);
-        /* Get the block header. */
-        word *hp = (word *) Hp_val (val);
-
-        while (Ecolor (*hp) == 0) hp = (word *) *hp;
-        CAMLassert (Ecolor (*hp) == 3);
-        if (Tag_ehd (*hp) == Closure_tag){
-          /* This is the first infix found in this block. */
-          /* Save original header. */
-          *p = *hp;
-          /* Link inverted infix list. */
-          Hd_val (q) = (header_t) ((word) p | 2);
-          /* Change block header's tag to Infix_tag, and change its size
-             to point to the infix list. */
-          *hp = Make_ehd (Wosize_bhsize (q - val), Infix_tag, 3, (uintnat) 0);
+  header_t h;
+
+  CAMLassert (((uintnat) p & 3) == 0);
+
+  if (Is_block (q)){
+    h = Hd_val (q);
+    switch (Color_hd (h)){
+    case Caml_white:
+      if (Tag_hd (h) == Infix_tag){
+        value realvalue = (value) q - Infix_offset_val (q);
+        if (Is_black_val (realvalue)){
+          break;
         }else{
-          CAMLassert (Tag_ehd (*hp) == Infix_tag);
-          /* Point the last of this infix list to the current first infix
-             list of the block. */
-          *p = (word) &Field (val, Wosize_ehd (*hp)) | 1;
-          /* Point the head of this infix list to the above. */
-          Hd_val (q) = (header_t) ((word) p | 2);
-          /* Change block header's size to point to this infix list. */
-          *hp = Make_ehd (Wosize_bhsize (q - val), Infix_tag, 3, (uintnat) 0);
+          CAMLassert (Is_in_heap (realvalue));
         }
+      }else{
+        CAMLassert (Is_in_heap (q));
       }
```

Comment:
That's a lot of code for a complicated assertion.  Remember that the fall through will assert `Is_in_heap(q)` again,
so the `else CAMLassert (Is_in_heap (q))` is redundant.  Also remember that `realvalue` and `q` are either both in the major heap or both outside, so the `CAMLassert (Is_in_heap (realvalue))` is redundant too.

What about:
```
  case Caml_white:

Reply:
Simplified.